File: Java/combinations.java
//  Combinations

// Time O(N!), Space O(N)
class Solution {
    private List<List<Integer>> ans = new ArrayList<List<Integer>>();
    private List<Integer> comb = new ArrayList<Integer>();
    
    public List<List<Integer>> combine(int n, int k) {
        dfs(n, k, 1);
        return ans;
    }
    
    private void dfs(int n, int k, int start) {
        if (comb.size() == k) {
            ans.add(new ArrayList(comb));
            return;
        }
        for (int i=start; i<=n; i++) {
            comb.add(i);
            dfs(n, k, i+1);
            comb.remove(comb.size()-1);
        }
    }
}

File: leetcode/combinations.cc
//  Combinations

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> comb;
        dfs(n, k, 1, comb, res);
        return res;
    }
    
    void dfs(int n, int k, int cur, vector<int> &comb, vector<vector<int>> &res) {
        if (k == 0) {
            res.push_back(comb);
            return;
        }
        for (int i=cur; i<=n; i++) {
            comb.push_back(i);
            dfs(n, k-1, i+1, comb, res);
            comb.pop_back();
        }
    }
};